akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/javadsl/Retry.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2024 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.dynamodb.javadsl
+
+import java.time.Duration
+import java.util.Optional
+import java.util.concurrent.CompletionStage
+import java.util.function.BiFunction
+import java.util.function.{ Function => JFunction }
+import java.util.{ List => JList }
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.jdk.CollectionConverters._
+import scala.jdk.DurationConverters._
+import scala.jdk.FutureConverters._
+import scala.jdk.OptionConverters._
+
+import akka.actor.typed.ActorSystem
+import akka.japi.function.Procedure3
+import akka.projection.dynamodb.scaladsl
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
+import software.amazon.awssdk.services.dynamodb.model.BatchWriteItemRequest
+import software.amazon.awssdk.services.dynamodb.model.BatchWriteItemResponse
+
+object Retry {
+
+  /**
+   * Perform a DynamoDB batch write, retrying writes for any unprocessed items (with exponential backoff).
+   *
+   * @param client the DynamoDB async client to use
+   * @param request the batch write request to apply
+   * @param maxRetries max retries, when reached the resulting future will be failed with `failOnMaxRetries`
+   * @param minBackoff minimum duration to backoff between retries
+   * @param maxBackoff maximum duration to backoff between retries
+   * @param randomFactor adds jitter to the retry delay (use 0 for no jitter)
+   * @param onRetry called before each retry, with the response, the current retry count, and the delay for this retry
+   * @param failOnMaxRetries if max retries is reached, create a throwable for the failed future result
+   * @param system the actor system (for scheduling)
+   * @return all responses from attempts (in order)
+   */
+  def batchWrite(
+      client: DynamoDbAsyncClient,
+      request: BatchWriteItemRequest,
+      maxRetries: Int,
+      minBackoff: Duration,
+      maxBackoff: Duration,
+      randomFactor: Double,
+      onRetry: Procedure3[BatchWriteItemResponse, Integer, Duration],
```

Comment:
What should this callback be used for, other than logging? shall we have a convenient variant without the callback parameter and log by default in that case? Maybe that convenience should also create the exception by default, so without `failOnMaxRetries` parameter.

Reply:
Yeah, we probably don't need the flexible callbacks for the convenience method at all. Will build in the logging and exception.